---
title: "Mapping report of BD Rhapsody TAS-Seq WTA data"
date: "`r Sys.Date()`"
author: "ImmunoGeneTeqs Inc."
output:
  rmdformats::readthedown:
    highlight: kate
    self_contained: true
    thumbnails: false
    lightbox: false
    gallery: false
    mathjax: false
    toc_depth: 2  #type-of-column header specification 
    md_extensions: -ascii_identifiers
params:
  samplename: samplename
  species: species
---

```{r setup, echo=FALSE, cache=FALSE, message=FALSE}
suppressMessages(library(knitr))
suppressMessages(library(rmdformats))
suppressMessages(library(data.table))
suppressMessages(library(dplyr))
suppressMessages(library(ggplot2))
suppressMessages(library(stringr))
suppressMessages(library(tibble))
suppressMessages(library(tidyr))
suppressMessages(library(DropletUtils))
suppressMessages(library(DT))
suppressMessages(library(cowplot))

quiet <- function(x) { 
  sink(tempfile()) 
  on.exit(sink()) 
  invisible(force(x)) 
} 

tableread_fast = function(i){
	 tmp = fread(i, header=FALSE, sep="\t", quote="")
   tmp = as.data.frame(tmp)
	 return(tmp)
}

## Global options
options(max.print="75")
opts_chunk$set(echo=FALSE,
	             cache=TRUE,
               prompt=FALSE,
               tidy=TRUE,
               comment=NA,
               message=FALSE,
               warning=FALSE)
opts_knit$set(width=50)
opts_template$set(
  fig_small = list(fig.width=5, fig.height=4, dpi=300, out.width = 200, out.height = 160), 
  fig_large = list(fig.width=5, fig.height=4, dpi=300, out.width = 500, out.height = 400)
  )

sample_name = params$samplename

if(params$species == "hsa"){
  reference = "GRCh38 release-101"
} else if (params$species == "mmu"){
  reference = "GRCm38 release-101"
} else if (params$species == "macaca"){
  reference = "Macaca_fascicularis_6.0 release-103"
} else if (params$species == "rat"){
  reference = "Rattus_norvegicus.Rnor_6.0 release-104"
} else if (params$species == "mermoset"){
  reference = "Callithrix_jacchus.ASM275486v1 release-104"
}
```

--------

## Base composition of adapter-trimmed/quality-filtered sequencing data

After adapter trimming and quanity filtering, base composition of sequencing data was analyzed by FastQC-v0.11.9.  

Figure 1a. Base compositoion of read1 data (cell barcodes). Note that the BD Rhapsody library contains fixed spacer sequences (position 10-21 and 31-43 base pair) and 96 types of 9-base known barcodes (position 1-9, 22-30, and 44-52 base pair).    

```{r, echo=FALSE, out.width = 600, out.height = 250, message=FALSE, fig.align='center', dpi=300}
# Define variable containing url
url1 = paste(sample_name, "_per_base_sequence_content_R1.png", sep="")
knitr::include_graphics(paste("../data/", sample_name,"_per_base_sequence_content_R1.png", sep=""))
```

Figure 1b. Base compositoion of read2 data (cDNA sequences). Note that the first few bases did not show uniform distribution becuase the sequencing library was generated by enzymatic fragmentation.  
```{r, echo=FALSE, out.width = 800, out.height = 250, message=FALSE, fig.align='center', dpi=300}
# Define variable containing url
url2 = paste(sample_name, "_per_base_sequence_content_R2.png", sep="")
knitr::include_graphics(paste("../data/", sample_name,"_per_base_sequence_content_R2.png", sep=""))
```

Base composition plot was stored at `r paste("./plots/", url1, sep="")` and `r paste("./plots/", url2, sep="")`.

--------

## Brief summary of the mapping pipeline procedures

Pair-end Fastq files (R1: cell barcode reads, R2: RNA reads) of sample `r paste(sample_name, sep="")` were processed as follows.  
Adapter trimming of sequencing data was performed by using Cutadapt 3.4.  
Filtered reads were chunked to 16 parts for parallel processing by using Seqkit 0.15.0.  
Filtered cell barcode reads were annotated by Python script provided by BD with minor modification for compatible to Python3.8.  
Associated cDNA reads were mapped to reference RNA (build `r paste(reference,sep="")`) by using bowtie2-2.4.2 by the following parameters: -p 2 --very-sensitive-local -N 1 --norc --seed 656565 --reorder.  
Then, cell barcode information of each read were added to the bowtie2-mapped BAM files, and read counts of each gene in each cell barcode were counted by using mawk.  
Resulted count data was converted to genes x cells matrix file and inflection threshold of the knee-plot was detected by using DropletUtils package in `r paste(version$version.string, sep="")`.  

```{r analysis, echo=FALSE, cache=FALSE}

#create output folders
setwd("../") #must set working directory. default is .Rmd-existing path.

#make output directories
dir.name = sprintf("./result/%s_results", sample_name)
if(!dir.exists(dir.name)){
dir.create(dir.name)
}
dir.name.1 = paste(dir.name, "/matrix", sep="")
if(!dir.exists(dir.name.1)){
dir.create(dir.name.1)
}
dir.name.1 = paste(dir.name, "/plots", sep="")
if(!dir.exists(dir.name.1)){
dir.create(dir.name.1)
}
dir.name.1 = paste(dir.name, "/stats", sep="")
if(!dir.exists(dir.name.1)){
 dir.create(dir.name.1)
}
dir.name.1 = paste(dir.name, "/processed", sep="")
if(!dir.exists(dir.name.1)){
 dir.create(dir.name.1)
}

#read count table, processed by data.table package
data1 = fread(paste("./result/", params$samplename,"_final_count.txt.gz", sep=""), header=FALSE, sep="\t", quote="")
colnames(data1)=c("Symbol", "barcode", "counts")

# calculate Total read counts and Detected Genes of each Cell-BCs by data.table package
tmp1 = data1[, by=barcode, sum(counts)] 
colnames(tmp1)=c("ALLCell_BCs", "Total_reads")
tmp2 = data1[, by=barcode, uniqueN(Symbol)]
colnames(tmp2)=c("ALLCell_BCs", "Symbol_kinds")
res2 = merge(tmp1, tmp2, by = "ALLCell_BCs", all = TRUE)
res2 = res2[order(-Total_reads)]
quiet(gc())

#select top50000 cell BCs by read counts
#avoid machine$integer.max error

if(length(unique(data1$Symbol))*50000 >= 2147483647){
 Cell_BCs = res2$ALLCell_BCs[1:as.integer(2147483000/length(unique(data1$Symbol)))]
} else {
  Cell_BCs = res2$ALLCell_BCs[1:50000]  
}
res3 = data1[barcode %chin% Cell_BCs]

#create gene x cell matrix
res4 = data.table::dcast(res3, Symbol ~ barcode, value.var = "counts", fill=0)
res4 = as.data.frame(res4)
tmp = res4[,1]
res4 = res4[,2:ncol(res4)]
rownames(res4)=tmp
file.name_top = sprintf("%s.txt.gz", sample_name)
file.name_top = paste("matrix_top", ncol(res4), "_", file.name_top, sep="")
file.name_top1 = paste(dir.name, "/matrix/", file.name_top, sep="")
fwrite(as.data.frame(res4), file.name_top1, row.names=T, col.names=T, 
       sep="\t", eol="\n", quote=F, compress="gzip")
symbol_total_50000 = nrow(res4)
quiet(gc())


#detect inflection point/knee-threshold
br.out = barcodeRanks(res4)
tmp = res4[,colSums(res4)>=10000] 
#print(paste("Survived cells : ", sample_name, "=", as.character(ncol(tmp))))
#print(paste("count_threshold:", sample_name, "=", as.character(10000)))
tmp1 = c(ncol(tmp), 10000)

#export cell x gene matrix file
#file.name = sprintf("./result/matrix_over10000_%s.txt", sample_name) 
#fwrite(as.data.frame(tmp), file.name, row.names=T, col.names=T, sep="\t", eol="\n", quote=F, nThread=32)

#define threshold as inflection point of knee-plot
tmp = res4[,colSums(res4)>=metadata(br.out)$inflection] 
#print(paste("Survived cells : ", sample_name, "=", as.character(ncol(tmp))))
#print(paste("inflection_threshold:", sample_name, "=", as.character(metadata(br.out)$inflection)))
tmp2 = c(ncol(tmp), metadata(br.out)$inflection)
symbol_total = nrow(tmp)

#export cell x gene matrix file
file.name_matrix = sprintf("matrix_inflection_%s.txt.gz", sample_name) 
file.name_matrix = paste(dir.name, "/matrix/", file.name_matrix, sep="")
fwrite(as.data.frame(tmp), file.name_matrix, row.names=T, col.names=T, sep="\t", 
       eol="\n", quote=F, nThread=32, compress="gzip")

#estimate Background beads by DropletUtils Package
#Background is picked up from under10000 cell barcodes in top50000 cell BCs
#set.seed(42)
#hoge1 = colSums(res4)
#hoge1 = hoge1[order(hoge1, decreasing=F)]
#res4 = as.matrix(res4)
#e.out = emptyDrops(res4, lower = max(hoge1[1:10000]), 
#                   retain = metadata(br.out)$inflection, 
#                   BPPARAM=BiocParallel::MulticoreParam(24), 
#                   niters=50000)

#is.cell = e.out$FDR <= 0.001
#e.out = as.data.frame(e.out)
#e.out$LogProb[is.na(e.out$LogProb)]=0
#e.out$LogProb = -e.out$LogProb
#e.out$PValue[is.na(e.out$PValue)]=1
#e.out$FDR[is.na(e.out$FDR)]=1
#colnames(e.out)[1]="Nreads"
#hoge2 = e.out$FDR
#hoge3 = hoge2 <= 0.001
#hoge4 = hoge2 > 0.001
#hoge2[hoge3]="Cell"
#hoge2[hoge4]="Background"
#e.out$Label = hoge2
#res5 = res4[,rownames(e.out[e.out$Label=="Cell",])]
#file.name_matrix = sprintf("matrix_DropletUtils_%s.txt.gz", sample_name) 
#file.name_matrix = paste(dir.name, "/matrix/", file.name_matrix, sep="")
#fwrite(as.data.frame(res5), file.name_matrix, row.names=T, col.names=T, sep="\t", 
#       eol="\n", quote=F, nThread=32, compress="gzip")

#export thresholds and survived cells
threshold_log = cbind(tmp1, tmp2)
rownames(threshold_log)=c("Survived_Cells", "count_threshold")
colnames(threshold_log)=c("count", "inflection")

#concatenate and reshape mapping statistics
res2=as.data.frame(res2)
res3 = res2[res2[,2]>=metadata(br.out)$inflection,]
read_stats = c(median(res3[,2]), mean(res3[,2]), sum(res3[,2]), sum(res2[,2]))
symbol_stats =c(median(res3[,3]), mean(res3[,3]), symbol_total, symbol_total_50000)
inflection_log = cbind(read_stats, symbol_stats)
rownames(inflection_log)=c("median", "mean", "total_inflection", "total")

cutadapt_log = sprintf("./result/%s_cutadapt.log", sample_name)
cutadapt_log = read.table(cutadapt_log, header=TRUE, row.names=1, sep="\t", 
                         quote="",stringsAsFactors=FALSE)
mapping_log = sprintf("./result/%s_mapping.log", sample_name)
mapping_log = read.table(mapping_log, header=TRUE, row.names=1, sep="\t", 
                         quote="",stringsAsFactors=FALSE)

cutadapt_log1 = cutadapt_log[,1]
mapping_log1 = mapping_log[,2]
threshold_log1 = threshold_log[1,1]
threshold_log2 = threshold_log[c(2,1),2]
inflection_log1 = inflection_log[c(3,1,2),1]
inflection_log2 = inflection_log[c(1:3),2]
mapping_log2 = inflection_log[4,1]/mapping_log[2,2]*100
mapping_log3 = inflection_log[3,1]/inflection_log[4,1]*100

logs = c(cutadapt_log1, mapping_log1, mapping_log2, 
         inflection_log[4,1], threshold_log2, threshold_log1, 
         inflection_log1[1], mapping_log3, inflection_log1[2:3], inflection_log2)
logs = as.data.frame(logs)
rownames(logs) = c("total reads", "trimmed reads", "mapped reads", 
                   "mapping rate against reference", 
                   "mapping rate against reference & CellBC", "final assigned reads",
                   "inflection threshold", "survived cells", "over10000read cells",
                   "total reads over inflection threshold", 
                   "percent of total reads over inflection threshold",
                   "median reads over inflection threshold", 
                   "mean reads over inflection threshold", 
                   "median symbols over inflection threshold", 
                   "mean symbols over inflection threshold", 
                   "total genes detected")
colnames(logs) = sample_name
logs=logs[c(1:8,10:16),,drop=F]
logs[c(4,5, 10,12,14),1] = round(logs[c(4,5, 10,12,14),1], digits=2)
logs[c(12,14),1] = round(logs[c(12,14),1], digits=0)
file.name.logs = sprintf("stats_concat_%s.txt", sample_name) 
file.name.logs = paste(dir.name, "/stats/", file.name.logs, sep="")
write.table(logs, file.name.logs, row.names=T, col.names=T, sep="\t", quote=F, append=F)

#export plots
o = order(br.out$rank)
df = data.frame(Rank=br.out$rank[o], 
                Nreads=br.out$total[o], 
                fitted=br.out$fitted[o])

df2 = data.frame(spline(df[,1:2]))
colnames(df2)=c("Rank", "Nreads")

p_knee = ggplot(df) + geom_point(aes(Rank, Nreads), size=0.5, colour="cyan") + 
  geom_line(data = df2, aes(Rank, Nreads), colour="cyan", size=1) +
  geom_vline(aes(xintercept=threshold_log[1,2], 
                 linetype=paste(threshold_log[1,2],"cells", sep=" ")), 
             colour="magenta", size=0.75) +
  scale_linetype_manual(name = "Inflection", values = c(1, 1), 
                        guide = guide_legend(override.aes = list(color = c("magenta"))))+
  scale_x_log10() +
  scale_y_log10() +
  ggtitle(paste("Knee-plot (survived cells = ", threshold_log[1,2]," cells)", sep="")) +
  theme_linedraw() +
  theme(plot.title=element_text(hjust = 0.5), text=element_text(size=8), legend.position = 'none')

file.name_knee = sprintf("counts_knee_inflection_%s.png", sample_name)
file.name_knee = paste(dir.name, "/plots/", file.name_knee, sep="")
ggsave(file=file.name_knee, plot=p_knee, device="png", units="in",
       dpi=300, width=5, height=4, limitsize = FALSE)

#plotting counts vs Detected genes over inflection threshold
res3 = res2[res2[,2]>=metadata(br.out)$inflection,]
df = data.frame(Reads=res3[,2], 
                Symbols=res3[,3])

p_count_symbol = ggplot(df,  aes(Reads, Symbols)) +
    geom_point(data=df, aes(Reads, Symbols), size=0.5, alpha=0.5, shape=20, colour="cyan") +
    geom_smooth(method="loess", aes(Reads, Symbols), colour="cyan", fill="cyan") +
    theme_linedraw() +
    ggtitle("Read counts versus symbols")+
    theme(plot.title=element_text(hjust = 0.5), 
          legend.position = 'none', text=element_text(size=8))
file.name = sprintf("CountvsSymbol_%s.png", sample_name)
file.name = paste(dir.name, "/plots/", file.name, sep="")
ggsave(file=file.name, plot=p_count_symbol, device="png", units="in",
       dpi=300, width=5.5, height=4, limitsize = FALSE)

```
---

## Results

### Mapping statistics

---

```{r output_table, echo=FALSE, cache=FALSE,dpi=300, out.width = 500, out.height = 500}
DT::datatable(logs, filter="none", width=700,
          caption = 'Table 1: mapping statistics, detected cell and detected gene statistics.', 
          class = 'cell-border stripe', 
          options = list(pageLength=16, searching=FALSE, lengthChange=FALSE)) %>%
   DT::formatStyle(columns = c(1, 2), fontSize = 8)
```

Original data is stored at `r sprintf("/stats/stats_concat_%s.txt", sample_name)`

---

### Knee-plot and read counts/gene symbols scatter plot

Figure 2. Knee-plot ((log10 read counts of each cell barcodes) x (rank of read counts of each cell barcodes)) and scatter plot (read count vs number of detected genes for each cell).

---

```{r output_plots1, echo=FALSE, cache=FALSE, fig.width=6, fig.height=2.5, dpi=300, out.width = 750, out.height = 300}

plot_grid(p_knee, p_count_symbol)

```


---


Plot file locations  

* Knee-plot was stored at `r sprintf("./plots/counts_knee_inflection_%s.png", sample_name)`.
* Scatter plot was stored at `r sprintf("./plots/CountvsSymbol_%s.png", sample_name)`.

---

## Output file explanations


Raw sequence files (fastq.gz format)  

* Raw sequence fastq.gz files were stored at `r paste("./fastq/", sep="")` folder. "R1" means read 1 sequence, and "R2" means read 2 sequence.  

Gene-expression matrix files (gzip compressed)

Rownames are gene symbols, and colnames are cell barcodes.

* Genes x cells matrix (over inflection threshod) was stored at `r sprintf("./matrix/matrix_inflection_%s.txt.gz", sample_name)`.
* Genes x cells matrix (top`r paste(ncol(res4), sep="")` cell barcodes) was stored at `r paste("./matrix/", file.name_top, sep="")`.

Processed count data (gzip comporessed, tidy data-format)  

* First row: gene symbols  
* Second column: cell barcodes  
* Third column: read counts  
    * processed count data was stored at `r sprintf("./processed/%s_final_count.txt.gz", sample_name)`.


---

## sessionInfo

```{r sessioninfo, results.folding="hide", summary.results="SessionInfo"}  
options(max.print = 999999)
sessionInfo()
```
